draft-jhoyla-req-mtls-flag.md

```diff
@@ -32,54 +32,73 @@ informative:
+them. In TLS 1.3 this may be done with post-handshake auth, however this adds an
+extra round-trip, and requires negotiation at the application layer.
 
 The behaviour specified in {{?RFC8446}} for a client that receives a
-CertificateRequest that it cannot satisfy is to send an empty Certificate
-message followed by a Finished message. However in practice many clients simply
+`CertificateRequest` that it cannot satisfy is to send an empty Certificate
+message followed by a `Finished` message. However in practice many clients simply
 terminate the connection in anticipation of a "certificate_required" alert.
 
-This behaviour makes it difficult for servers to send CertificateRequest
-messages in the wild. A client sending the request mTLS flag in the ClientHello
+This behaviour makes it difficult for servers to send `CertificateRequest`
+messages in the wild. A client sending the `request_mtls` flag in the `ClientHello`
 allows the server to request authentication during the initial handshake only
 when it receives a hint the client will handle the request without terminating
 the connection, and, if unable to authenticate, by sending an empty certificate
 message, per {{?RFC8446, Section 4.4.2}}. This enables a number of use cases,
 for example allowing bots to authenticate themselves when mixed in with general
 traffic.
 
+This flag is intended to be used by clients sending automated traffic, not those
+that are directly operated by (human) users. Clients that set this flag SHOULD
+only do so when a client certificate is directly available to them, and MUST NOT
+prompt the user or take actions that would prompt the user for example trying to
+read certificates from off-board devices such as smart cards.
+
 # Conventions and Definitions
 
 {::boilerplate bcp14-tagged}
 
 # Flag specification
 
-A server receiving this flag MAY send a CertificateRequest message, and a client
-sending this flag MUST send a Certificate message if it receives a
-CertificateRequest, even if said Certificate message is empty, and SHOULD NOT
-preemptively terminate the connection anticipating a "certificate_required"
-alert.
+The `request_mtls` flag is sent by the client in the `ClientHello` message,
+and, if accepted by the server, acknowledged in the server's
+`CertificateRequest` message. A server that receives this flag and wishes to
+accept the extension MUST send a `CertificateRequest` message with the TLS
+Flags extension, and with the `request_mtls` flag set. The server MAY send a
+`CertificateRequest` message without the `request_mtls` flag set if it wishes
+to negotiate client authentication for another reason. The client sending this
+flag MUST send a `Certificate` message if it receives a `CertificateRequest`
+with the `request_mtls` flag set, even if said `Certificate` message is empty,
+and SHOULD NOT preemptively terminate the connection anticipating a
+"certificate_required" alert.
```

Comment:
The SHOULD NOT fills me with pause. As I understand it, the goal of this flag is to express to the server that the client will handle the CR in a specific way: respond with an (empty) CT. Is there a legitimate reason why a client might ignore this SHOULD NOT?

Reply:
Yes. If the server sets the `certificate_authorities` extension, and only lists authorities that the client doesn't support, ditto with `signature_algorithms`. Also, the client may wish to terminate the connection for another reason (e.g. server cert didn't verify), so we shouldn't require them to continue.